src/ai/backend/manager/models/alembic/versions/1d42c726d8a3_add_container_registry_tables.py

```diff
@@ -0,0 +1,304 @@
+
+    db_connection = op.get_bind()
+    rows = db_connection.execute(sa.select(ContainerRegistryRow)).fetchall()
+    items = []
+
+    for id, url, registry_name, type, project, username, password, ssl_verify, _is_global in rows:
+        item = {
+            "": url,
+            "type": str(type),
+            "hostname": registry_name,  # registry_name should be reverted to hostname,
+        }
+
+        if project is not None:
+            item["project"] = project
+        if username is not None:
+            item["username"] = username
+        if password is not None:
+            item["password"] = password
+        if ssl_verify is not None:
+            item["ssl_verify"] = ssl_verify
+
+        items.append(item)
+
+    # Records with the same hostname should be grouped together.
+    # Note that if there are columns other than the project column that have different values between records,
+    # information loss can occur.
+    grouped_items = {k: list(v) for k, v in groupby(items, key=lambda x: x["hostname"])}
+
+    def merge_items(items):
+        for item in items:
+            if "project" not in item:
+                return items[0]
+
+        projects = [item["project"] for item in items]
+        merged_projects = ",".join(projects)
+
+        merged_item = items[0].copy()
+        merged_item["project"] = merged_projects
+
+        return merged_item
+
+    merged_items = [merge_items(items) for items in grouped_items.values()]
+
+    def put_etcd_container_registries(merged_items: list[Any], queue: Queue):
+        etcd = get_async_etcd()
+        for item in merged_items:
+            hostname = item.pop("hostname")
+            asyncio.run(etcd.put_prefix(f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}", item))
+
+        queue.put(True)
+
+    queue: Queue = Queue()
+
+    with ThreadPoolExecutor() as executor:
+        executor.submit(put_etcd_container_registries, merged_items, queue)
+
+    queue.get()
+
+
+def insert_registry_id_to_images() -> None:
+    db_connection = op.get_bind()
+    ContainerRegistry = get_container_registry_row_schema()
+
+    image_infos = db_connection.execute(sa.select([ImageRow.id, ImageRow.name])).fetchall()
+
+    for image_info in image_infos:
+        image_id, image_name = image_info
+
+        parts = image_name.split("/")
+        registry_name = parts[0]
+        project = parts[1]
+
+        registry_id = db_connection.execute(
+            sa.select(ContainerRegistry.id)
+            .where(ContainerRegistry.registry_name == registry_name)
+            .where(ContainerRegistry.project == project)
+        ).scalar()
+
+        if registry_id is not None:
+            db_connection.execute(
+                sa.update(ImageRow)
+                .values(registry_id=str(registry_id))
+                .where(ImageRow.id == image_id)
+            )
+        else:
+            print(f"ContainerRegistry row not found for image {image_name}", file=sys.stderr)
```

Comment:
Let's iterate all `ContainerRegistries` and save their `id`. Such as `registry_name/project` key and `id` value dictionary.
Then we can bulk update image rows here.
Below is an example bulk update using `bindparam`. you can try and let me know how it works!
```python
stmt = (
    sa.update(ImageRow)
    .where(
        (ImageRow.registry_name == sa.bindparam("registry_name"))
        & (ImageRow.project == sa.bindparam("project"))
    )
    .values(ImageRow.registry_id == sa.bindparam("registry_id"))
)
value = [
{
    "registry_name": d["registry_name"],
    "project": d["project"],
    "registry_id: d["registry_id"],
} for d in data
]
db_connection.execute()
```